### ISSUE TITLE
More output series or decoder wrappers?

### ISSUE THREAD

**Issue by jlibovicky** [ufal/neuralmonkey#59]
Do we want to be able to get more outputs from a single decoder? The use cases why we may want to have something like this are:
- be able to get attention while decoding
- do something as factored translation (e.g., get a pair of POS and lemma from a decoder)

There are basically two options how we can achieve that. One is to allow the decoder to have multiple output series. Then we would need to specify in the trainer, which series it should work with. The other option (if we think having more output series would be a rare case and assuming we have already implemented multiple decoders), is to have decoder wrapper, that would get out the other output from the original decoder.

I would probably go for the second one, although it may obfuscate the design a bit. On the other hand, it will keep the implementation simple and keep the property of having one output per decoder. It will leave only one issue to be solved: what to with the trainer(s) in case of multiple decoders.

What do you think?

**Comment by jindrahelcl** [comment]
I personally like more the second approach. It puts the middle-man decoders to somewhat similar position as encoders, but with input not from datasets but from other encoders.

If we could abstract from decoders and encoders and design an interface how to connect these objects, we would still have a nice, modular and simple design, and we could make use of the advantages it brings.

Back to the issue - if there will be only one top-level decoder, then what is the problem with the trainer?

**Comment by jlibovicky** [comment]
What is according to you the second approach: multiple series in one decoder?

The problem with trainer is the same in both cases: what if one series should be optimized using cross-entropy, the second one would be a regression with L2 loss and the third one would not have any reference output (e.g., attention). The current trainer is indeed not capable of doing that. The solution might be to separate loss computation and feed it to a one big trainer? We might have different ways of computing loss for the same decoder (e.g. MIXER or minimum risk training) and if the loss is complicated, we can benefit from not constructing its graph at runtime.

**Comment by jlibovicky** [comment]
Well, after a short mountain run, I have the following suggestion: 

1. keep cross-entropy trainer almost as it is (+ we would need to specify series ID) ... for keeping the configuration simple in simple cases
2. create module `objectives` where the objective functions will be implemented - cross-entropy, MIXER, minimum risk training (an objective would get a decoder, a series ID and return a objective)
3. create a `MultiTrainer` (or whatever we will call it) that will have the objectives and an argument

This would solve both the problem of multiple decoder and multiple series in one decoder. Any commets? Hate speeches?

**Comment by jlibovicky** [comment]
I mean API, of course, there's no need to implement it twice.

I am not sure either and it depends on who our users will be. I suppose, one group of users may be people who don't know much about deep learning, have some data and want to try out how the model would work with their data. Such users may appreciate to have such syntactic sugar. (Or may not, I don't know). Until now, a trainer with exactly one decoder producing exactly one series was enough and I guess it will be in most of the cases, it would be meaningful to have shortcut for that.

**Comment by jlibovicky** [comment]
Solved with #151.
